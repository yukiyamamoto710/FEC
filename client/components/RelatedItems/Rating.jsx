import React from 'react';
import Stars from '../Reviews/Stars/Stars';
import PropTypes from 'prop-types';

const Rating = ({rating}) => {
  var rates = Object.keys(rating).sort();
  var size = rates.reduce((sum, i)=>(sum + Number(rating[i])),0);
  var total = rates.reduce((sum, i)=>(sum + Number(i) * Number(rating[i])), 0);
  var ave  = total/size;

  return (
    <div>
     <Stars rate={ave} name={'name'} classNameForSize={'msgStar'}/>
    </div>
  )
}

Rating.propTypes = {
  rating: PropTypes.object
}

export default Rating;
