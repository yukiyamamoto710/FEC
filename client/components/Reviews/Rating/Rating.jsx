import React from 'react';
import PropTypes from 'prop-types';
import StarList from './StarList/StarList';
import Characteristics from './Characteristics/Characteristics';
import { Data } from '../data';
import Stars from '../Stars/Stars';

const Rating = (props) => {
  const { rating, starClicked } = props;
  const obj = { ...rating.ratings };
  const key = Object.keys(obj).sort();
  let num = 0;
  const keyArray = [...Array(5)].map(() => {
    num += 1;
    return String(num);
  });
  const valueArray = keyArray.map((i) => {
    if (obj[i] !== undefined) {
      return Number(obj[i]);
    }
    return 0;
  });
  const size = key.reduce((sum, i) => (sum + Number(obj[i])), 0);
  const perArray = valueArray.map((i) => Math.floor((i / size) * 100));
  const total = key.reduce((sum, i) => (sum + Number(i) * (Number(obj[i]))), 0);
  let ave = 0;
  if (size !== 0) {
    ave = Math.round((total / size) * 10) / 10;
  }
  let tNum = 0;
  if (rating.recommended.true) {
    tNum = Number(rating.recommended.true);
  }
  let fNum = 0;
  if (rating.recommended.false) {
    fNum = Number(rating.recommended.false);
  }
  let rAve = 0;
  if (tNum + fNum !== 0) {
    rAve = Math.floor((tNum / (tNum + fNum)) * 100);
  }
  let cKey = [];
  if (rating.characteristics !== undefined) {
    cKey = Object.keys(rating.characteristics);
  }
  return (
    <div
      className="rating"
      data-testid="rating"
    >
      <div
        className="ratingHead"
        data-testid="head"
      >
        {ave}
        <Stars
          rate={ave}
          classNameForSize="ratingStar"
        />
      </div>
      <h5 data-testid="text">
        {`${rAve}% of reviews recommend this product`}
      </h5>
      <div data-testid="starListR">
        <StarList
          stars={keyArray.reverse()}
          starClicked={starClicked}
          perArray={perArray}
        />
      </div>
      {cKey.map((i) => (
        <div
          key={rating.characteristics[i].id}
          data-testid="characteristics"
        >
          <Characteristics
            dataArray={Data[i]}
            name={i}
            value={Number(rating.characteristics[i].value)}
          />
        </div>
      ))}
    </div>
  );
};

Rating.propTypes = {
  rating: PropTypes.shape({
    product_id: PropTypes.string,
    ratings: PropTypes.objectOf(PropTypes.string),
    recommended: PropTypes.objectOf(PropTypes.string),
    characteristics: PropTypes.objectOf(
      PropTypes.shape({
        id: PropTypes.number,
        value: PropTypes.string,
      }),
    ),
  }),
  starClicked: PropTypes.func,
};

Rating.defaultProps = {
  rating: {},
  starClicked: () => 1,
};
export default Rating;
