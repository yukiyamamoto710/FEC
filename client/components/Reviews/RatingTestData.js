const testData1 = {
  characteristics: {
    Comfort: {
      id: 86687,
      value: '2.7500000000000000',
    },
    Quality: {
      id: 86688,
      value: '2.0000000000000000',
    },
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
    Width: {
      id: 86686,
      value: '3.2500000000000000',
    },
  },
  product_id: '25821',
  ratings: {
    1: '1',
    4: '1',
    5: '2',
  },
  recommended: {
    false: '1',
    true: '3',
  },
};
const testData2 = {
  characteristics: {
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
  },
  product_id: '25721',
  ratings: {
  },
  recommended: {
  },
};

export { testData1, testData2 };
