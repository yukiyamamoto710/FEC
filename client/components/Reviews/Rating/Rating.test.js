/**
 * @jest-environment jsdom
 */
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  cleanup,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Rating from './Rating';
import { testData1, testData2 } from '../RatingTestData';

afterEach(cleanup);
describe('render correct', () => {
  it('render Rating with testData1', () => {
    const { getByTestId } = render(<Rating rating={testData1} />);
    expect(getByTestId('rating')).toBeInTheDocument();
  });
  it('render Rating with testData2', () => {
    const { getByTestId } = render(<Rating rating={testData2} />);
    expect(getByTestId('rating')).toBeInTheDocument();
  });
  it('head should not be NaN, testData1', () => {
    const { getByTestId } = render(<Rating rating={testData1} />);
    expect(getByTestId('head')).not.toHaveTextContent('NaN');
  });
  it('head should not be NaN, testData2', () => {
    const { getByTestId } = render(<Rating rating={testData2} />);
    expect(getByTestId('head')).not.toHaveTextContent('NaN');
  });
  it('text should not have NaN%, testData1', () => {
    const { getByTestId } = render(<Rating rating={testData1} />);
    expect(getByTestId('text')).toHaveTextContent('75% of reviews recommend this product');
  });
  it('text should not have NaN%, testData2', () => {
    const { getByTestId } = render(<Rating rating={testData2} />);
    expect(getByTestId('text')).toHaveTextContent('0% of reviews recommend this product');
  });
  it('starList should always be 5, testData2', () => {
    const { getAllByTestId } = render(<Rating rating={testData2} />);
    expect(getAllByTestId('starListR').length).toBe(1);
  });
  it('starList should always be 5, testData1', () => {
    const { getAllByTestId } = render(<Rating rating={testData1} />);
    expect(getAllByTestId('starListR').length).toBe(1);
  });
  it('characteristics should be same amount as keys, testData1', () => {
    const { getAllByTestId } = render(<Rating rating={testData1} />);
    const keys = Object.keys(testData1.characteristics);
    expect(getAllByTestId('characteristics')).toHaveLength(keys.length);
  });
  it('characteristics should be same amount as keys, testData2', () => {
    const { getAllByTestId } = render(<Rating rating={testData2} />);
    const keys = Object.keys(testData2.characteristics);
    expect(getAllByTestId('characteristics')).toHaveLength(keys.length);
  });
});

it('should have default starsClicked()', () => {
  expect(Rating.defaultProps.starClicked()).toBeDefined();
});

it('matches snapShot', () => {
  const tree = renderer.create(<Rating rating={testData1} />).toJSON();
  expect(tree).toMatchSnapshot();
});
