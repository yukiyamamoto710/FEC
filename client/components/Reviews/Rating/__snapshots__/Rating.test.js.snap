// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<div
  className="rating"
  data-testid="rating"
>
  <div
    className="ratingHead"
    data-testid="head"
  >
    3.8
    <div
      className="ratingHead"
      data-testid="stars"
    >
      <div>
        <div
          className="ratingStar fullStar"
          data-testid="button0"
          id=" 1"
          label="fullStar"
          name=""
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        />
      </div>
      <div>
        <div
          className="ratingStar fullStar"
          data-testid="button1"
          id=" 2"
          label="fullStar"
          name=""
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        />
      </div>
      <div>
        <div
          className="ratingStar fullStar"
          data-testid="button2"
          id=" 3"
          label="fullStar"
          name=""
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        />
      </div>
      <div>
        <div
          className="ratingStar qThreeStar"
          data-testid="button3"
          id=" 4"
          label="qThreeStar"
          name=""
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        />
      </div>
      <div>
        <div
          className="ratingStar emptyStar"
          data-testid="button4"
          id=" 5"
          label="emptyStar"
          name=""
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        />
      </div>
    </div>
  </div>
  <h5
    data-testid="text"
  >
    75% of reviews recommend this product
  </h5>
  <div
    data-testid="starListR"
  >
    <div
      className="starListContainer"
      data-testid="starList"
    >
      <div>
        <div
          className="starListBase"
        >
          <button
            className="starListText"
            data-testid="startListButton5"
            id="5"
            onClick={[Function]}
            type="button"
          >
            5 Stars
          </button>
          <div
            className="starRatingBarBase"
            data-testid="starRatingBar"
          >
            <div
              style={
                Object {
                  "backgroundColor": "rgb(102, 162, 182)",
                  "border": "5px solid rgb(102, 162, 182)",
                  "width": "50%",
                }
              }
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="starListBase"
        >
          <button
            className="starListText"
            data-testid="startListButton4"
            id="4"
            onClick={[Function]}
            type="button"
          >
            4 Stars
          </button>
          <div
            className="starRatingBarBase"
            data-testid="starRatingBar"
          >
            <div
              style={
                Object {
                  "backgroundColor": "rgb(102, 162, 182)",
                  "border": "5px solid rgb(102, 162, 182)",
                  "width": "25%",
                }
              }
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="starListBase"
        >
          <button
            className="starListText"
            data-testid="startListButton3"
            id="3"
            onClick={[Function]}
            type="button"
          >
            3 Stars
          </button>
          <div
            className="starRatingBarBase"
            data-testid="starRatingBar"
          >
            <div
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="starListBase"
        >
          <button
            className="starListText"
            data-testid="startListButton2"
            id="2"
            onClick={[Function]}
            type="button"
          >
            2 Stars
          </button>
          <div
            className="starRatingBarBase"
            data-testid="starRatingBar"
          >
            <div
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="starListBase"
        >
          <button
            className="starListText"
            data-testid="startListButton1"
            id="1"
            onClick={[Function]}
            type="button"
          >
            1 Stars
          </button>
          <div
            className="starRatingBarBase"
            data-testid="starRatingBar"
          >
            <div
              style={
                Object {
                  "backgroundColor": "rgb(102, 162, 182)",
                  "border": "5px solid rgb(102, 162, 182)",
                  "width": "25%",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="characteristics"
  >
    <div
      className="charaContainer"
      data-testid="chara"
    >
      Comfort
      <div
        className="charaBar"
      >
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          style={
            Object {
              "bottom": "10%",
              "left": "55%",
              "position": "absolute",
            }
          }
        >
          <div
            className="charaBarPoint"
          />
        </div>
      </div>
      <div
        className="barTextContainer"
      >
        <div
          className="barText"
        >
          Uncomfortable
        </div>
        <div
          className="barText"
        >
          Perfect
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="characteristics"
  >
    <div
      className="charaContainer"
      data-testid="chara"
    >
      Quality
      <div
        className="charaBar"
      >
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          style={
            Object {
              "bottom": "10%",
              "left": "40%",
              "position": "absolute",
            }
          }
        >
          <div
            className="charaBarPoint"
          />
        </div>
      </div>
      <div
        className="barTextContainer"
      >
        <div
          className="barText"
        >
          Poor
        </div>
        <div
          className="barText"
        >
          Perfect
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="characteristics"
  >
    <div
      className="charaContainer"
      data-testid="chara"
    >
      Size
      <div
        className="charaBar"
      >
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          style={
            Object {
              "bottom": "10%",
              "left": "45%",
              "position": "absolute",
            }
          }
        >
          <div
            className="charaBarPoint"
          />
        </div>
      </div>
      <div
        className="barTextContainer"
      >
        <div
          className="barText"
        >
          A size too small
        </div>
        <div
          className="barText"
        >
          A size too wide
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="characteristics"
  >
    <div
      className="charaContainer"
      data-testid="chara"
    >
      Width
      <div
        className="charaBar"
      >
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          className="charaSmallBar"
        />
        <div
          style={
            Object {
              "bottom": "10%",
              "left": "65%",
              "position": "absolute",
            }
          }
        >
          <div
            className="charaBarPoint"
          />
        </div>
      </div>
      <div
        className="barTextContainer"
      >
        <div
          className="barText"
        >
          Too narrow
        </div>
        <div
          className="barText"
        >
          Too wide
        </div>
      </div>
    </div>
  </div>
</div>
`;
